### ISSUE TITLE
v8: Forms Image upload not working when using FormPrePopulate

### ISSUE THREAD

**Issue by ArnoldV** [umbraco/Umbraco.Forms.Issues#177]
(Goldpartner: Axendo)

In a default form setup on Umbraco v8 the file upload field does not seem to post any file to the filesystem.

The default path ("~/app_data/TEMP/FileUploads") stays empty.

We have the idea that 'ProcessSubmittedValue' is called when requesting the form, but not when submitting the form when using the 'Umbraco.Forms.Web.Controllers.UmbracoFormsController.FormPrePopulate += UmbracoFormsController_FormPrePopulate". But without the possiblity do debug agains source this is a big guess as why this is happening.

steps:

1. Umbraco v8 (cloud) 
2. Umbraco Forms form with default upload field 
3. Add form to page (using the grid) 
4. visit the page 
5. Add file to upload field 
6. Add a "Umbraco.Forms.Web.Controllers.UmbracoFormsController.FormPrePopulate += UmbracoFormsController_FormPrePopulate" handler to the site.
7. Submit the form

nothing happens with the file.

Also when having a workflow that sends an email and have 'attached uploaded files', does not add any files to the mail as well.

We also tried the UmbracoForms.8.1.1-gamma.10104.nupkg nightly.

---
_This item has been added to our backlog [AB#2324](https://umbraco.visualstudio.com/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/2324)_

**Comment by MvdDonk** [comment]
@ArnoldV 
It appears this problem occurs when an event is attached to UmbracoFormsController.FormPrePopulate. Removing `Umbraco.Forms.Web.Controllers.UmbracoFormsController.FormPrePopulate += UmbracoFormsController_FormPrePopulate;` solves this issue above. 

Adding it will make the upload stop working even if there isn't any logica in "UmbracoFormsController_FormPrePopulate"

**Comment by kbtan-uda** [comment]
Hi @nul800sebastiaan,

Can I suggest and present a code fix here?

Regards,
Kwok

**Comment by nul800sebastiaan** [comment]
@kbtan-uda Of course, don't see why not! 👍

**Comment by kbtan-uda** [comment]
In UmbracoFormsController class, ExtractAllPagesState function, add code to get the upload file submited details and save it to form state.
```
private Dictionary<string, object[]> ExtractAllPagesState(FormViewModel model, ControllerContext context, Form form)
        {
            var currentState = RetrieveFormState(model);

            if (currentState == null)
            {
                return null;
            }

            foreach (var field in form.AllFields)
            {
                object[] finalValues = new object[0];


                var prePopValues = form.AllFields.First(f => f.Id == field.Id).Values == null ?
                    new object[0] : form.AllFields.First(f => f.Id == field.Id).Values.ToArray();


                if (context.HttpContext.Request.Form.AllKeys.Contains(field.Id.ToString()))
                {
                    var postedvalues = context.HttpContext.Request.Form.GetValues(field.Id.ToString());

                    bool isEmpty = true;
                    if (postedvalues != null)
                    {
                        foreach (var postedvalue in postedvalues)
                        {
                            if (string.IsNullOrEmpty(postedvalue) == false)
                            {
                                isEmpty = false;
                                break;
                            }
                        }

                        // If the postedValues is empty, it means that the pre-populated field has been
                        // manually cleared and it can be pre-populated, otherwise keep the values as it is, we do this as
                        // the postedValues can hold a value that has been pre-populated or entered by a user, and because
                        // it could be a user input we always keep it, unless is empty.
                        finalValues = isEmpty ? prePopValues : postedvalues;
                    }
                }
                else if (context.HttpContext.Request.Files.AllKeys.Contains(field.Id.ToString()))
                {
                    var fieldSchema = form
                                .AllFields
                                .First(f => f.Id == field.Id);

                    finalValues = fieldSchema.FieldType
                        .ProcessSubmittedValue(fieldSchema, finalValues, context.HttpContext).ToArray();
                }
                else
                {
                    finalValues = prePopValues;
                }

                if (currentState.ContainsKey(field.Id.ToString()))
                    currentState[field.Id.ToString()] = finalValues;
                else
                    currentState.Add(field.Id.ToString(), finalValues);
            }
```

**Comment by stefankip** [comment]
Running into this issue as well with umbraco v7.15.4. Major showstopper...

**Comment by markwemekamp** [comment]
Any progress on this issue? We're running into this (Umbraco v 7.15.6 with Forms 7.4.1) aswell.

Reproducable by creating a form with a file upload and introducing an ApplicationEventHandler with an empty FormPrePopulate event

**Comment by AndyButland** [comment]
Resolved for upcoming V8 release 8.7.3,

**Comment by lucvansoest** [comment]
> Resolved for upcoming V8 release 8.7.3,

Is there also a fix scheduled for V7? Running in the same issue and migrating to V8 is not a viable option for us at the moment.

**Comment by ronaldbarendse** [comment]
@stefankip and @lucvansoest I've re-opened #229 to track the process for v7.
